internal/states/statefile/version4.go

```diff
@@ -139,9 +145,22 @@ func prepareStateV4(sV4 *stateV4) (*File, tfdiags.Diagnostics) {
 
 			instAddr := rAddr.Instance(key)
 
+			// The provider can either be set on the resource level or on the instance level
+			var instanceProviderAddr addrs.AbsProviderConfig
+			if isV4.InstanceProvider != "" {
```

Comment:
Let's also handle somehow (panic?) the case when provider is not set anywhere. This should never happen, but still better to handle it as early as possible.

Reply:
https://github.com/opentofu/opentofu/pull/1963/commits/1182ee49506a9cd491ca9f276b2ca784293a19b9